Add support for debug f-string

msm-code opened the pull request:
I have a feeling that you won't like it @rocky, but...

I've added support for debug f-strings by moving the complexity to post-processing. The code now "just" parses debug f-strings like normal f-strings, and looks for patterns of `XYZ={XYZ!r}` (and similar). I did this because I found the previous version quite buggy - for example it didn't support  `f'a{a=}b'`, `f'{abc=}{a}'`, `f'{abc=}a{a}'`, etc. To support arbitrary number of expressions in f-string it had to be generalised anyway (I think).

Another problem was that python combines string literals, so for example `a{a=}` will be compiled to `LOAD_STR aa=` and `LOAD_VALUE a`. Again, i've handled that during post-processing in semantic action.

Finally, if that approach is acceptable, I think "normal" and debug f-strings should be unified. Currently `f'{f}{g}{abc=}{a}{b}{c}{d}{e}'` is parsed using the "normal" fstring grammar, and because of this it decompiles to `abc={abc!r}` instead of `{abc=}` - not wrong, but suboptimal.

I'm happy to clean up this code and fix the rough edges, but submitting PR for (initial) review in case my approach is completely wrong and it should be done in some completely different way.

Fixes #105.

**edit** oh, and one more thing - I didn't find `test_debug_conversion` method in `test_fstring.py`, and `3.8-fstring-debug-attribute` seems to be entirely merged into master. Is #105's description correct?

rocky commented:
@msm-code I just looked briefly at this and it doesn't bother me at all on first look. It will be a little while before I can look at this in detail which I am anxious to do.

Note that there is something like this going on for list comprehensions. Looking a this purely from a code standpoint you have some code calling an inner function with a parameter which in Python is something like `.0`. If you look at pycdc output you'll sometime see list comprehensions this way as a function with that `.0` parameter. 

But in the uncompyle-derived code there has always been a bit of effort to try to stitch thing back together as a comprehension.   It  to more complexity in the code base, but I guess it is a good thing rather than punt and describe the code in a way that you really can't use in running back in Python.

rocky commented:
Just had a change to look at in more depth. Generally, I am pretty fussy, but I don't have any problem with this. (The example in the comment 

>         # Caveat: python merges strings on bytecode level, so "a{a=}" becomes "aa={a!r}"

I found confusing, rather than something more expansive like: 

>  "user setting: {user=}" becomes "settings: user={user!r}"

but this is all within a marge on acceptability.

rocky commented:
CI seems to say that you have more examples in mind.

msm-code commented:
Great. In this case I'll clean this up, and submit a improved version.

I'm going to have some time over the weekend so probably I'll finish it soon (and maybe pick up another ticket).

> CI seems to say that you have more examples in mind that haven't been worked out.

About the CI fault: I've looked briefly at https://github.com/rocky/python-decompile3/actions/runs/4179906825/jobs/7295573884, and the error is raised by third praty library import on this line (not added by me):

```
test_fstring.py:12: in <module>
    from hypothesis import strategies as st
```

I had the same problem locally, and I've solved it locally by bumping hypothesis:

```diff
❯ git diff requirements.txt                                                                                                                                                                                   ncspot
diff --git a/requirements.txt b/requirements.txt
index 62771f9c..dcd66ab1 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,5 +1,5 @@
 # Pick up stuff from setup.py
-hypothesis==2.0.0
+hypothesis==6.24.2
 pytest~=7.1.3
```

I didn't commit it because I've figured that's an unrelated change, and probably a problem with my system. I'll add it to the next version of my PR.

rocky commented:
> About the CI fault: I've looked briefly at https://github.com/rocky/python-decompile3/actions/runs/4179906825/jobs/7295573884, and the error is raised by third party library import on this line (not added by me):
> 
> ```
> test_fstring.py:12: in <module>
>     from hypothesis import strategies as st
> ```
> 
> I had the same problem locally, and I've solved it locally by bumping hypothesis:
> 
> ```diff
> ❯ git diff requirements.txt                                                                                                                                                                                   ncspot
> diff --git a/requirements.txt b/requirements.txt
> index 62771f9c..dcd66ab1 100644
> --- a/requirements.txt
> +++ b/requirements.txt
> @@ -1,5 +1,5 @@
>  # Pick up stuff from setup.py
> -hypothesis==2.0.0
> +hypothesis==6.24.2
>  pytest~=7.1.3
> ```
> 
> I didn't commit it because I've figured that's an unrelated change, and probably a problem with my system. I'll add it to the next version of my PR.
> 

Thanks for the explanation and fix. I have made this change in 31b6ce97

msm-code commented:
Squashed my changes and rebased them onto the current master (to benefit from https://github.com/rocky/python-decompile3/commit/31b6ce9777ff5696d734e5b8e8b8772388bc3d18).

I've cleaned up and simplified the code a bit. I've also removed the edge case where strings of form `{a}_{b=}` were parsed with older grammar rules and didn't benefit from this PR. But overall no major functional changes in comparison to the version you've reviewed. 

I can't trigger workflows to check if the tests pass, but running `pytest` in my local directory does work:

```
msm@transient ~/Projects/python-decompile3/pytest on  feature/debug-fstrings [?] via 🐍 v3.8.16 (venv)
❯ pytest
================================================================================================ test session starts ================================================================================================
platform linux -- Python 3.8.16, pytest-7.1.3, pluggy-1.0.0
rootdir: /home/msm/Projects/python-decompile3
plugins: hypothesis-6.24.2
collected 33 items

test_basic.py ..                                                                                                                                                                                              [  6%]
test_build_const_key_map.py ..........                                                                                                                                                                        [ 36%]
test_code_deparse.py ...                                                                                                                                                                                      [ 45%]
test_deparse.py .                                                                                                                                                                                             [ 48%]
test_fstring.py .............                                                                                                                                                                                 [ 87%]
test_grammar.py ..                                                                                                                                                                                            [ 93%]
test_lambda_grammar.py .                                                                                                                                                                                      [ 96%]
test_token.py .                                                                                                                                                                                               [100%]
```

As far as I'm concerned the code is ready to merge (but do re-check).

rocky commented on decompyle3/semantics/customize38.py:
```diff
@@ -303,41 +303,53 @@ def n_formatted_value_debug(node):
+        # Collect all expressions that we're formatting one by one
+        # Fragments is a list of parsed expressions, like ['a', '{a=}', 'b']
+        # Caveat: python merges strings on bytecode level, so for example
+        # f"user setting: {user=}" becomes f"settings: user={user!r}".
```
A small typo which is my mistake `f"settings: ` -> `f"setting: `

msm-code replied:
Right. Missed that. Fixed.

msm-code commented:
Looks like the tests still fail.

I've looked at the pipeline configuration, and looks like it installs requirements from `requirements-dev.txt`:

```
.github/workflows/osx.yml:        pip install -r requirements-dev.txt
.github/workflows/ubuntu.yml:        pip install -r requirements-dev.txt
.github/workflows/windows.yml:        pip install -r requirements-dev.txt
```

and there's still old hypothesis version there:

```
❯ grep -R "hypothesis<="
requirements-dev.txt:hypothesis<=3.0.0
```

Shall I bump the version there too, or maybe do this in a separate PR?

rocky commented:
> Looks like the tests still fail.
> 
> I've looked at the pipeline configuration, and looks like it installs requirements from `requirements-dev.txt`:
> 
> ```
> .github/workflows/osx.yml:        pip install -r requirements-dev.txt
> .github/workflows/ubuntu.yml:        pip install -r requirements-dev.txt
> .github/workflows/windows.yml:        pip install -r requirements-dev.txt
> ```
> 
> and there's still old hypothesis version there:
> 
> ```
> ❯ grep -R "hypothesis<="
> requirements-dev.txt:hypothesis<=3.0.0
> ```
> 
> Shall I bump the version there too, or maybe do this in a separate PR?

I think/hope I have this addressed with https://github.com/rocky/python-decompile3/commit/eac77e82c54652255fe160eab809da7bf27bf528

triggering a rebuild of the failed jobs though doesn't seem to pick this up. Would you try rebasing (yet again) and let's see if we can get this to pass. 

If it is easier and you want I can just add you as a developer to this project.

msm-code commented:
> triggering a rebuild of the failed jobs though doesn't seem to pick this up. Would you try rebasing (yet again) and let's see if we can get this to pass.

Sure, done.

> If it is easier and you want I can just add you as a developer to this project.

To auto-trigger the pipelines? I think the default configuration (if you didn't change is) that pipelines are only restricted for first time contributor, so it should get easier when I get my first PR merged.

Looks like there's a real test fail now. I'll investigate. (Maybe I should try to get a test env more similar to the github workflow, it definitely passes locally)

msm-code commented:
Ok, that should do it - I didn't handle nested formats correctly. I've only ran `pytest` and missed out on all the thorough tests ran by `make check`.

rocky commented:
YAY!  Thanks. 

BTW, there is more to do just on the f-string side. Somewhere in trying to understand what f= does, I see that there can be now comments inside f-strings. 

Also, the way I discovered that there was a problem in the first place was by trying to decompile `test_fstring.py` which comes with the 3.8+ distribution. See `test/stdlib/3.8-exclude.sh`

rocky commented:
@msm-code if you want a _real_ torture test for fstrings cd into `test/stdlib` make sure you are using Python 3.8 or later and run: 

```
$ ./runtests3.sh test_fstring.py
Python version is 3.8.16
'.python-version' -> '../.python-version'
'.python-version' -> '../../.python-version'
compiling test_fstring.py to test_fstring-3.8.pyc
========== 09:51:44 AM Decompiling test_fstring.py ===========
========== 09:51:48 AM Running test_fstring.py ===========
  File "test_fstring.py", line 11
    --
     ^
SyntaxError: invalid syntax
** Ran 1 tests before failure. Skipped 0 test for known failures. **

```

I think this _also_ fails for 3.7 before the `f=` assignments.

rocky commented:
I have had this in the back of my mind for a while so I'll say it here. 

First, I think it pretty awesome that we now have f-string assignment specifier working now. I suspect that _none_ of the other Python decompilers support this.  And of course you or others are free to adapt this so that they do.

So when someone comes to the discussion as says I'll offer $100 for a decompiler but _only_ if it is a perfect decompiler, this to me is silly across a number of dimensions. (I don't go to a doctor and say I'll give you $100 but only if you cure me.) 

But that silliness aside, the point here in just the work alone is to show that there is an _incredible_ amount of work needed to write a complete Python decompiler. 

Sure, you may say, hey, no one uses f= assignments in 3.8, so you can largely leave that out. That may be true now, but what has happened is that over time as people become aware that this exists, it is used in 3.9, 3.10, etc. and then even in 3.8.  Type annotations were like this. They were around in some form in 3.4 and earlier, but only as 3.6 kicked in did they get used more. And anyone today using 3.4 for whatever reason is more likely to use type annotations than when 3.4 first came out. 

Also, something on the order of writing a new version for the compiler itself. Maybe less than, but still more "on the order of".

And while there are people who have full-time jobs writing Python compilers, I am not aware of anyone who full time write Python decompilers or has gotten paid for that in conjunction with there work. Well, there have been occasional contributors with someone anonymous backgrounds, so I don't know if that is true, but aside from you, no one has fessed up with saying that they use this at work.

msm-code commented:
> @msm-code if you want a real torture test for fstrings cd into test/stdlib make sure you are using Python 3.8 or later and run:

Still waiting for the pyenv to finish building before I respond in more details. Anyway in the meantime:

> BTW, there is more to do just on the f-string side. Somewhere in trying to understand what f= does, I see that there can be now comments inside f-strings.

Not sure this is the root issue, internet says comments inside f-strings are not allowed (at least in 3.8):

https://realpython.com/python-f-strings/

I confirm that they don't work locally for me too:

```
>>> f"Eric is {2 * 37 #Oh my!}."
  File "<stdin>", line 1
SyntaxError: f-string expression part cannot include '#'
```

> So when someone comes to the discussion as says I'll offer $100 for a decompiler but only if it is a perfect decompiler, this to me is silly across a number of dimensions. (I don't go to a doctor and say I'll give you $100 but only if you cure me.)

Sure, if this was a commercial project, this is an houry consulting rate in many parts of the world (and judging by your github profile, not even 30 minutes of your time). I hope nobody seriously expects $100 is a significant price for **building a working decompiler**.

> I am not aware of anyone who full time write Python decompilers or has gotten paid for that in conjunction with there work. Well, there have been occasional contributors with someone anonymous backgrounds, so I don't know if that is true, but aside from you, no one has fessed up with saying that they use this at work.

I don't get paid for working on decompilers too :upside_down_face:, but I know a lot of people who use them at work. Actually my whole industry does (malware reverse engineering). Granted, reverse-engineers don't necessarily make good engineers, so many couldn't contribute even if they wanted to. There are probably many more other legitimate work uses of the tool that I'm not aware.

rocky commented:
> Granted, reverse-engineers don't necessarily make good engineers, so many couldn't contribute even if they wanted to. 

Most programmer who use compilers couldn't to the compiler either. Yet somehow compiler development seems important enough that it is usually funded and not done on a volunteer basis. 

> There are probably many more other legitimate work uses of the tool that I'm not aware.

Suppose your service uses 3rd party Python libraries that are not open source and you want to ensure that no sensitive data of your leaks out through this service? 

My initial interest was getting more accurate information about where the program is when stopped in a debugger or when an exception occurred.  Those are two, that are not hypothetical examples, but real.

rocky commented:
@msm-code for the present, I've reverted this. When `{{` is fixed we should try merging again. 

(I will be making some small type-cleanup commits soon.)
